File: packages/shared/src/answerQuality.ts
/**
 * Deterministic quality checks for a live "Say aloud" answer. Used by regression
 * tests (and available to the Test Lab) to assert answers are spoken-ready:
 * within length, no ChatGPT tails, no leaked internal labels, a direct opening.
 *
 * This is heuristic and language-aware (Russian live answers) — it does not judge
 * factual correctness, only the structural "does it sound like a candidate" rules.
 */

/** Phrases that must never appear in a live/Say-aloud answer. */
export const FORBIDDEN_LIVE_PHRASES: string[] = [
  'если хотите, могу подробнее',
  'если хотите, могу разложить',
  'важно отметить',
  'в заключение',
  'давайте рассмотрим',
  'main answer',
  'key points',
  'short answer',
  'я не совсем понял вопрос. если говорить в общем',
  'в разных контекстах могут быть разные подходы',
  'это позволило мне углубить',
  'существуют различные инструменты и методы',
];

/** Internal section labels that should be stripped before display. */
export const INTERNAL_LABELS: string[] = [
  'main answer',
  'key points',
  'short answer',
  'detailed',
];

/** Filler openings that make the first sentence non-direct. */
const FILLER_OPENINGS = [
  /^похоже,/i,
  /^вероятно,/i,
  /^судя\s+по\s+всему,/i,
  /^я\s+понял\s+вопрос\s+как/i,
  /^вопрос\s+(?:про|касается)/i,
  /^можно\s+сказать,/i,
  /^в\s+целом,/i,
  /^давайте\s+разберём/i,
  /^важно\s+отметить/i,
  /^я\s+не\s+совсем\s+понял/i,
];

export interface SpokenAnswerQuality {
  wordCount: number;
  withinWordLimit: boolean;
  forbiddenPhrases: string[];
  internalLabels: string[];
  startsWithFiller: boolean;
  hasMarkdownHeader: boolean;
  /** Overall: spoken-ready (within limit, clean, direct opening). */
  ok: boolean;
}

export function countWords(text: string): number {
  const t = (text || '').trim();
  return t ? t.split(/\s+/).length : 0;
}

export function scoreSpokenAnswer(
  answer: string,
  opts: { maxWords?: number; requiredTerms?: string[] } = {},
): SpokenAnswerQuality {
  const maxWords = opts.maxWords ?? 90;
  const text = (answer || '').trim();
  const low = text.toLowerCase();

  const forbiddenPhrases = FORBIDDEN_LIVE_PHRASES.filter((p) => low.includes(p));
  const internalLabels = INTERNAL_LABELS.filter((l) =>
    new RegExp(`(?:^|\\n|[.?!]\\s)\\s*\\*{0,2}\\s*${l}\\s*\\*{0,2}\\s*:`, 'i').test(low),
  );
  const wordCount = countWords(text);
  const withinWordLimit = wordCount <= maxWords;
  const startsWithFiller = FILLER_OPENINGS.some((re) => re.test(text));
  const hasMarkdownHeader = /(?:^|\n)\s*#{1,6}\s/.test(text);

  const ok =
    wordCount > 0 &&
    withinWordLimit &&
    forbiddenPhrases.length === 0 &&
    internalLabels.length === 0 &&
    !startsWithFiller &&
    !hasMarkdownHeader;

  return {
    wordCount,
    withinWordLimit,
    forbiddenPhrases,
    internalLabels,
    startsWithFiller,
    hasMarkdownHeader,
    ok,
  };
}

File: apps/api-py/app/services/answer_quality.py
"""Deterministic quality scoring for a live Say-aloud answer (Python side).

Mirrors packages/shared/src/answerQuality.ts. Used by the offline answer-quality
eval harness (scripts/eval_answers.py) to score real LLM answers, and unit-tested
on crafted good/bad answers. It judges structure/say-aloud-ness, not factual
correctness — required-term presence is the caller's job via `required_terms`.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

# Phrases that must never appear in a live / Say-aloud answer.
FORBIDDEN_LIVE_PHRASES = [
    "если хотите, могу подробнее",
    "если хотите, могу разложить",
    "важно отметить",
    "в заключение",
    "давайте рассмотрим",
    "main answer",
    "key points",
    "short answer",
    "я не совсем понял вопрос. если говорить в общем",
    "в разных контекстах могут быть разные подходы",
    "это позволило мне углубить",
    "существуют различные инструменты и методы",
]

INTERNAL_LABELS = ["main answer", "key points", "short answer", "detailed"]

_FILLER_OPENINGS = [
    re.compile(r"^похоже,", re.IGNORECASE),
    re.compile(r"^вероятно,", re.IGNORECASE),
    re.compile(r"^судя\s+по\s+всему,", re.IGNORECASE),
    re.compile(r"^я\s+понял\s+вопрос\s+как", re.IGNORECASE),
    re.compile(r"^вопрос\s+(?:про|касается)", re.IGNORECASE),
    re.compile(r"^можно\s+сказать,", re.IGNORECASE),
    re.compile(r"^в\s+целом,", re.IGNORECASE),
    re.compile(r"^давайте\s+разберём", re.IGNORECASE),
    re.compile(r"^важно\s+отметить", re.IGNORECASE),
    re.compile(r"^я\s+не\s+совсем\s+понял", re.IGNORECASE),
]


@dataclass
class SpokenAnswerQuality:
    word_count: int
    within_word_limit: bool
    forbidden_phrases: list[str] = field(default_factory=list)
    internal_labels: list[str] = field(default_factory=list)
    missing_terms: list[str] = field(default_factory=list)
    starts_with_filler: bool = False
    has_markdown_header: bool = False
    ok: bool = False


def count_words(text: str) -> int:
    t = (text or "").strip()
    return len(t.split()) if t else 0


def score_spoken_answer(
    answer: str,
    *,
    max_words: int = 90,
    required_terms: list[str] | None = None,
) -> SpokenAnswerQuality:
    text = (answer or "").strip()
    low = text.lower()

    forbidden = [p for p in FORBIDDEN_LIVE_PHRASES if p in low]
    labels = [
        label
        for label in INTERNAL_LABELS
        if re.search(rf"(?:^|\n|[.?!]\s)\s*\*{{0,2}}\s*{re.escape(label)}\s*\*{{0,2}}\s*:", low)
    ]
    word_count = count_words(text)
    within = word_count <= max_words
    starts_filler = any(p.search(text) for p in _FILLER_OPENINGS)
    has_header = bool(re.search(r"(?:^|\n)\s*#{1,6}\s", text))
    missing = [t for t in (required_terms or []) if t.lower() not in low]

    ok = (
        word_count > 0
        and within
        and not forbidden
        and not labels
        and not starts_filler
        and not has_header
        and not missing
    )
    return SpokenAnswerQuality(
        word_count=word_count,
        within_word_limit=within,
        forbidden_phrases=forbidden,
        internal_labels=labels,
        missing_terms=missing,
        starts_with_filler=starts_filler,
        has_markdown_header=has_header,
        ok=ok,
    )
